flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java

```diff
@@ -0,0 +1,317 @@
+        // we need to delay the execution because this may trigger a state transition of the
+        // AdaptiveScheduler.
+        context.getMainThreadExecutor()
+                .execute(
+                        () ->
+                                stopWithSavepointOperationManager.abortOperation(
+                                        new CheckpointException(
+                                                "Aborting stop with savepoint operation for transition to state "
+                                                        + newState.getSimpleName()
+                                                        + ".",
+                                                CheckpointFailureReason.CHECKPOINT_ABORTED)));
+
+        super.onLeave(newState);
+    }
+
+    @Override
+    public void cancel() {
+        context.goToCanceling(
+                getExecutionGraph(), getExecutionGraphHandler(), getOperatorCoordinatorHandler());
+    }
+
+    @Override
+    public JobStatus getJobStatus() {
+        return JobStatus.RUNNING;
+    }
+
+    @Override
+    public void handleGlobalFailure(Throwable cause) {
+        handleAnyFailure(cause);
+    }
+
+    /**
+     * The {@code executionTerminationsFuture} will complete if a task reached a terminal state, and
+     * {@link StopWithSavepointOperationManager} will act accordingly.
+     */
+    @Override
+    boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionStateTransition) {
+        final boolean successfulUpdate =
+                getExecutionGraph().updateState(taskExecutionStateTransition);
+
+        if (successfulUpdate) {
+            if (taskExecutionStateTransition.getExecutionState() == ExecutionState.FAILED) {
+                Throwable cause = taskExecutionStateTransition.getError(userCodeClassLoader);
+                handleAnyFailure(cause);
+            }
+        }
+
+        return successfulUpdate;
+    }
+
+    @Override
+    void onGloballyTerminalState(JobStatus globallyTerminalState) {
+        context.goToFinished(ArchivedExecutionGraph.createFrom(getExecutionGraph()));
+    }
+
+    private void handleAnyFailure(Throwable cause) {
+        final Executing.FailureResult failureResult = context.howToHandleFailure(cause);
+
+        if (failureResult.canRestart()) {
+            context.goToRestarting(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    failureResult.getBackoffTime());
+        } else {
+            context.goToFailing(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    failureResult.getFailureCause());
+        }
+    }
+
+    CompletableFuture<String> getOperationCompletionFuture() {
+        return operationCompletionFuture;
+    }
+
+    @Override
+    public void startCheckpointScheduler() {
```

Comment:
I think this method shouldn't be executed unconditionally. What if we are no longer in the state `StopWithSavepoint`? Then a callback from the `StopWithSavepointOperationImpl` should not have any effect.